File: LeetCode/Problems/Python/543. Diameter of Binary Tree.py
#DFS
#Time: O(n)
#SpaceL O(n)
class Solution:
    def dfs(self, node):
        if not node:
            return 0
        
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        self.ans = max(self.ans, right+left+1)
        return max(left,right)+1
        
    
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        self.ans = float(-inf)
        self.dfs(root)
        
        #ans is the count of node, and we want to return the count of edge
        return self.ans-1

File: LeetCode/Problems/C++/543. Diameter of Binary Tree.cpp
class Solution {
private:
    int dfs(TreeNode* root, int& diameter){
        if(!root) return 0;

        int l = dfs(root->left, diameter);
        int r = dfs(root->right, diameter);
        diameter = max(l+r, diameter);
        
        return max(l, r)+1;
    }
    
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        int diameter =INT_MIN;
        dfs(root, diameter);
        

        return diameter;
    }
};